### PR TITLE
[DDP] Make _ReplicateState inherit from _State and make replicate eagerly initialized

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #109647
* #110688

Follow how fully_shard store the _FSDPState, this PR makes _ReplicateState inherit from _State.  This PR also makes replicate eagerly initialize the internal DDP instance so that users can access the required methods/functions before the first forward().

Differential Revision: [D49428291](https://our.internmc.facebook.com/intern/diff/D49428291/)

### CODE REVIEW

File: torch/distributed/_composable/replicate.py [hunk 1]
```diff
@@ -11,59 +12,30 @@
...
-
-def _is_fully_sharded(module: nn.Module) -> bool:
-    r"""Check if module is marked with fully_shard."""
-    return "fully_shard" in _get_registry(module)
-
-
-class _ReplicateState:
-    def __init__(self, ignored_modules: Optional[Iterable[torch.nn.Module]]) -> None:
+class _ReplicateState(_State):
+    def __init__(self) -> None:
```
Comment (wconstab) [inline torch/distributed/_composable/replicate.py:16]: no call to super()? 

what's the reason to subclass _State

Comment (wz337) [reply torch/distributed/_composable/replicate.py:16]: I believe the reason is that we are using the _State for composable APIs. Correct me if I am wrong.

File: torch/distributed/_composable/replicate.py [hunk 2]
```diff
@@ -11,59 +12,30 @@
...
+        self.ignored_modules: Set[torch.nn.Module] = set()
+        self.ignored_params: Set[torch.nn.Parameter] = set()
         # Only used for testing
         self._param_names: List[str] = []
 
-    def mark_module(self, module: nn.Module, **kwargs) -> None:
+    def mark_module(
+        self,
+        module: nn.Module,
+        ignored_modules: Optional[Iterable[torch.nn.Module]],
```
Comment (wz337) [inline torch/distributed/_composable/replicate.py:29]: QQ: Should `ignored_params` also be explicitly added to the signature?

**Review by wz337 — APPROVED** [review]
LGTM

File: torch/distributed/_composable/replicate.py (hunk 2 above)
Comment (fegin) [reply torch/distributed/_composable/replicate.py:30]: Since the `ignored_modules` is Optional for `replicate`, I think we can leave it as Optional now.

**Review by rohan-varma — APPROVED** [review]

**Pull request closed without merging**